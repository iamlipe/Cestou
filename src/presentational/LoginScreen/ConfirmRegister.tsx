import React, {useEffect} from 'react';
import styled from 'styled-components/native';
import {t} from 'i18next';
import {RouteProp, useNavigation, useRoute} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {AuthStackParamList} from '@/routes/stacks/AuthStack';

import Producer from '@/assets/svgs/welcome-producer.svg';
import Consumer from '@/assets/svgs/welcome-consumer.svg';

import Button from '@/components/Button';

type NavProps = NativeStackNavigationProp<AuthStackParamList, 'Onboarding'>;

type ParamList = {
  params: {
    userType: 'producer' | 'consumer';
    phoneOrEmail: string;
    password: string;
  };
};

export const ConfirmRegister = () => {
  const {navigate, addListener} = useNavigation<NavProps>();
  const route = useRoute<RouteProp<ParamList, 'params'>>();

  useEffect(() => {
    addListener('beforeRemove', event => {
      event.preventDefault();
    });
  });

  return (
    <StyledContainerScroll
      contentContainerStyle={{alignItems: 'center'}}
      showsVerticalScrollIndicator={false}>
      {route.params.userType === 'producer' ? (
        <StyledProducer testID="welcome-producer" />
      ) : (
        <StyledConsumer testID="welcome-consumer" />
      )}
      <StyledTitle>
        {t('text.screenConfirmRegister.registrationSuccessfully')}
      </StyledTitle>
      <StyledSubtitle>
        {t('text.screenConfirmRegister.validateYourRegistration')}
      </StyledSubtitle>
      <StyledConfirmButton
        testID="confirm-button"
        title={t('button.start')}
        onPress={() =>
          navigate('Onboarding', {
            phoneOrEmail: route.params.phoneOrEmail,
            password: route.params.password,
          })
        }
      />
    </StyledContainerScroll>
  );
};

const StyledContainerScroll = styled.ScrollView`
  min-height: 100%;
  background-color: ${({theme}) => theme.colors.BACKGROUND};
`;

const StyledProducer = styled(Producer)`
  margin-top: 40px;
`;

const StyledConsumer = styled(Consumer)`
  margin-top: 40px;
`;

const StyledTitle = styled.Text`
  font-family: ${({theme}) => theme.fonts.SEMIBOLD_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.LARGER};
  color: ${({theme}) => theme.colors.PRIMARY_800};

  margin-top: 40px;
`;

const StyledSubtitle = styled.Text`
  width: 90%;
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  text-align: center;
  color: ${({theme}) => theme.colors.GRAY_900};

  margin-top: 40px;
`;

const StyledConfirmButton = styled(Button)`
  margin-top: 40px;
`;
