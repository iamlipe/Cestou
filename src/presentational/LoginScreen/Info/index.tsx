import React, {memo} from 'react';
import styled from 'styled-components/native';
import {Linking} from 'react-native';
import {OUR_SITE} from 'react-native-dotenv';
import {useTranslation} from 'react-i18next';

import Button from '@/components/Button';

const Info = () => {
  const {t} = useTranslation();

  return (
    <StyledContainerInfo>
      <StyledText>{t('text.componentInfo.title')}</StyledText>
      <Button
        onPress={async () => Linking.openURL(OUR_SITE)}
        size="small"
        buttonColor="text_only"
        textColor="primary"
        title={t('button.goToSite')}
        noMargin
      />
    </StyledContainerInfo>
  );
};

const StyledContainerInfo = styled.View`
  width: 90%;
  align-self: center;
  justify-content: space-between;
  align-items: center;
  border: 1px solid ${({theme}) => theme.colors.PRIMARY_800};
  border-radius: 5px;
  padding: 8px;
  margin-bottom: 50px;
`;

const StyledText = styled.Text`
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
  font-size: ${({theme}) => theme.sizing.SMALLER};
  color: ${({theme}) => theme.colors.GRAY_900};
  margin-bottom: 16px;
`;

export default memo(Info);
