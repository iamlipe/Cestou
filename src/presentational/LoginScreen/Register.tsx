import React from 'react';
import styled from 'styled-components/native';

export const Register = () => {
  return (
    <StyledContainer>
      <StyledTitle>Register</StyledTitle>
    </StyledContainer>
  );
};

const StyledContainer = styled.SafeAreaView``;
const StyledTitle = styled.Text``;
