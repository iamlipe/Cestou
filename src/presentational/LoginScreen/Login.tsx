import React from 'react';
import styled from 'styled-components/native';
import * as Yup from 'yup';
import {Keyboard} from 'react-native';
import {SubmitHandler, useForm} from 'react-hook-form';
import {LOGIN, LoginRequest} from '@/store/slices/userSlice';
import {yupResolver} from '@hookform/resolvers/yup';
import {useReduxDispatch} from '@/hooks/useReduxDispatch';
import {useReduxSelector} from '@/hooks/useReduxSelector';
import {useNavigation} from '@react-navigation/native';
import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {RoutesParamList} from '@/routes';

import Logo from '@/assets/svgs/logo.svg';

import InputForm from '@/components/InputForm/index';
import Button from '@/components/Button';
import CheckboxForm from '@/components/CheckboxForm';
import Info from '@/components/Info';

const schema = Yup.object().shape({
  phoneOrEmail: Yup.string()
    .email('Informe um email válido')
    .required('Preenchimento obrigatório'),
  password: Yup.string().required('Preenchimento obrigatório'),
});

type NavProps = NativeStackNavigationProp<RoutesParamList, 'Register'>;

export const Login = () => {
  const userReducer = useReduxSelector(({user}) => user);
  const {navigate} = useNavigation<NavProps>();

  const dispatch = useReduxDispatch();
  const {
    control,
    handleSubmit,
    formState: {errors, isSubmitted},
  } = useForm<LoginRequest>({
    resolver: yupResolver(schema),
  });

  const onSubmit: SubmitHandler<LoginRequest> = async data => {
    dispatch(
      LOGIN({
        phoneOrEmail: data.phoneOrEmail,
        password: data.password,
      }),
    );

    Keyboard.dismiss();
  };

  return (
    <StyledContainer>
      <StyledContainerScroll showsVerticalScrollIndicator={false}>
        <StyledLogo testID="logo" />
        <StyledContainerForm>
          <InputForm
            control={control}
            name="phoneOrEmail"
            label="E-mail"
            error={
              (isSubmitted && errors.phoneOrEmail?.message) ||
              (userReducer.error?.response && 'telefone ou senha incorretos')
            }
          />
          <InputForm
            control={control}
            name="password"
            label="Senha"
            secureTextEntry
            error={
              (isSubmitted && errors.password?.message) ||
              (userReducer.error?.response && 'telefone ou senha incorretos')
            }
          />

          <StyledRowRecoverPassword>
            <StyledText>Esqueceu a senha?</StyledText>
            <StyledLink
              buttonColor="text_only"
              textColor="primary"
              title="Recuperar senha"
              size="medium"
              noMargin
              onPress={() => null}
            />
          </StyledRowRecoverPassword>

          <CheckboxForm
            control={control}
            name="remember_me"
            options={['Lembrar minha senha']}
          />
        </StyledContainerForm>
        <StyledButtonSubmit
          onPress={handleSubmit(onSubmit)}
          loading={userReducer.isLoading}
          title="Entrar"
          noMargin
        />

        <StyledRowRegister>
          <StyledText>Ainda não tem conta?</StyledText>
          <StyledLink
            buttonColor="text_only"
            textColor="primary"
            title="Cadastre-se"
            noMargin
            onPress={() => navigate('Register')}
          />
        </StyledRowRegister>
        <StyledLine />
        <Info />
      </StyledContainerScroll>
    </StyledContainer>
  );
};

const StyledContainer = styled.SafeAreaView``;

const StyledContainerScroll = styled.ScrollView`
  background-color: ${({theme}) => theme.colors.BACKGROUND};
`;

const StyledContainerForm = styled.View`
  width: 90%;
  align-self: center;
`;

const StyledLogo = styled(Logo)`
  align-self: center;
  margin: 64px 0;
`;

const StyledRowRecoverPassword = styled.View`
  flex-direction: row;
  align-items: flex-start;
  justify-content: space-between;
  margin-bottom: 20px;
`;

const StyledRowRegister = styled(StyledRowRecoverPassword)`
  align-self: center;
  margin-top: 20px;
`;

const StyledLink = styled(Button)`
  width: auto;
  height: auto;
  margin: 0 5px;
`;

const StyledLine = styled.View`
  width: 90%;
  height: 1px;
  align-self: center;
  background-color: ${({theme}) => theme.colors.GRAY_100};
  margin-bottom: 20px;
`;

const StyledButtonSubmit = styled(Button)`
  margin-top: 40px;
`;

const StyledText = styled.Text`
  font-family: ${({theme}) => theme.fonts.REGULAR_SOURCESANSPRO};
`;
