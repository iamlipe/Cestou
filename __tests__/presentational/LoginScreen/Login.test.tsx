import React from 'react';
import {act, fireEvent} from '@testing-library/react-native';
import {
  renderWithThemeProvider,
  matchSnapshotWithProvider,
} from '@/config/functions/testing';

import {Login} from '@/presentational/LoginScreen/Login';

describe('Login', () => {
  test('should render screen correctly', () => {
    renderWithThemeProvider(<Login />);
  });

  test('should matches snapshot', () => {
    matchSnapshotWithProvider(<Login />);
  });

  test('submit form', async () => {
    const {getByTestId, getByText} = renderWithThemeProvider(<Login />);

    const emailInput = getByTestId('input-text-phoneOrEmail');
    const passwordInput = getByTestId('input-text-password');

    fireEvent.changeText(emailInput, 'test@email.com');
    fireEvent.changeText(passwordInput, 'passowrd');

    await act(() => fireEvent.press(getByText(/entrar/i)));

    expect(emailInput.props.value).toEqual('test@email.com');
    expect(passwordInput.props.value).toEqual('passowrd');
  });

  test('should render Logo', () => {
    const {getByTestId} = renderWithThemeProvider(<Login />);

    const logo = getByTestId('logo');

    expect(logo).toBeTruthy();
  });
});
