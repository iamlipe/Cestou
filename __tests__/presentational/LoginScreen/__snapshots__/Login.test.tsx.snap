// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should matches snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {},
    ]
  }
>
  <RCTScrollView
    showsVerticalScrollIndicator={false}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
        },
      ]
    }
  >
    <View>
      <SvgMock
        style={
          Array [
            Object {
              "alignSelf": "center",
              "marginBottom": 64,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 64,
            },
          ]
        }
        testID="logo"
      />
      <View
        style={
          Array [
            Object {
              "alignSelf": "center",
              "width": "90%",
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "marginBottom": 10,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#f3fbf5",
                  "borderBottomWidth": 1,
                  "borderColor": "#3A3A3A",
                  "flexDirection": "row",
                  "height": 56,
                  "paddingBottom": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 0,
                  "width": "100%",
                },
              ]
            }
          >
            <Text
              animatedStyle={
                Object {
                  "value": Object {
                    "fontSize": 14,
                    "opacity": 1,
                    "transform": Array [
                      Object {
                        "translateY": 0,
                      },
                    ],
                  },
                }
              }
              collapsable={false}
              style={
                Object {
                  "color": "#1A1A1A",
                  "fontSize": 14,
                  "opacity": 1,
                  "paddingLeft": 16,
                  "position": "absolute",
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                }
              }
            >
              E-mail
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              secureTextEntry={false}
              style={
                Array [
                  Object {
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontSize": 27,
                    "marginTop": 15,
                  },
                ]
              }
              testID="input-text-phoneOrEmail"
            />
          </View>
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "flex-end",
                  "color": "#781B1C",
                  "fontFamily": "SourceSansPro-Regular",
                  "fontSize": 24,
                  "marginBottom": 3,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "marginBottom": 10,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#f3fbf5",
                  "borderBottomWidth": 1,
                  "borderColor": "#3A3A3A",
                  "flexDirection": "row",
                  "height": 56,
                  "paddingBottom": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 0,
                  "width": "100%",
                },
              ]
            }
          >
            <Text
              animatedStyle={
                Object {
                  "value": Object {
                    "fontSize": 14,
                    "opacity": 1,
                    "transform": Array [
                      Object {
                        "translateY": 0,
                      },
                    ],
                  },
                }
              }
              collapsable={false}
              style={
                Object {
                  "color": "#1A1A1A",
                  "fontSize": 14,
                  "opacity": 1,
                  "paddingLeft": 16,
                  "position": "absolute",
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                }
              }
            >
              Senha
            </Text>
            <TextInput
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontSize": 27,
                    "marginTop": 15,
                  },
                ]
              }
              testID="input-text-password"
            />
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "overflow": "hidden",
                  },
                  Object {
                    "backgroundColor": "#f3fbf5",
                    "borderRadius": 5,
                  },
                ]
              }
              testID="visibility-icon"
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                      "padding": 8,
                    },
                    Object {
                      "backgroundColor": "#f3fbf5",
                      "borderRadius": 5,
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    Array [
                      Object {
                        "color": "#1A1A1A",
                        "fontSize": 20,
                      },
                      Array [
                        Object {
                          "marginRight": 10,
                        },
                        Object {
                          "marginRight": 0,
                        },
                      ],
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                >
                  î£´
                </Text>
              </View>
            </View>
          </View>
          <Text
            style={
              Array [
                Object {
                  "alignSelf": "flex-end",
                  "color": "#781B1C",
                  "fontFamily": "SourceSansPro-Regular",
                  "fontSize": 24,
                  "marginBottom": 3,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "flex-start",
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginBottom": 20,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "fontFamily": "SourceSansPro-Regular",
                },
              ]
            }
          >
            Esqueceu a senha?
          </Text>
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "center",
                "backgroundColor": "transparent",
                "borderRadius": 4,
                "flexDirection": "row",
                "height": "auto",
                "justifyContent": "center",
                "marginBottom": 0,
                "marginLeft": 5,
                "marginRight": 5,
                "marginTop": 0,
                "opacity": 1,
                "width": "auto",
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#00843F",
                    "fontFamily": "SourceSansPro-SemiBold",
                    "fontSize": 27,
                  },
                ]
              }
              textColor="primary"
            >
              Recuperar senha
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginBottom": 10,
              "opacity": 1,
            }
          }
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#00843F",
                  "fontSize": 24,
                },
                undefined,
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
            testID="checkbox-icon-not-checked-0"
          >
            î µ
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#1A1A1A",
                  "fontFamily": "SourceSansPro-Regular",
                  "fontSize": 27,
                  "marginLeft": 5,
                },
              ]
            }
          >
            Lembrar minha senha
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "center",
            "backgroundColor": "#00843F",
            "borderRadius": 4,
            "flexDirection": "row",
            "height": 49,
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 40,
            "opacity": 1,
            "width": 200,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontFamily": "SourceSansPro-SemiBold",
                "fontSize": 27,
              },
            ]
          }
          textColor="white"
        >
          Entrar
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "flex-start",
              "alignSelf": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 20,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro-Regular",
              },
            ]
          }
        >
          Ainda nÃ£o tem conta?
        </Text>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "transparent",
              "borderRadius": 4,
              "flexDirection": "row",
              "height": "auto",
              "justifyContent": "center",
              "marginBottom": 0,
              "marginLeft": 5,
              "marginRight": 5,
              "marginTop": 0,
              "opacity": 1,
              "width": "auto",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#00843F",
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 27,
                },
              ]
            }
            textColor="primary"
          >
            Cadastre-se
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "borderColor": "#00843F",
              "borderRadius": 5,
              "borderStyle": "solid",
              "borderWidth": 1,
              "justifyContent": "space-between",
              "marginBottom": 50,
              "paddingBottom": 24,
              "paddingLeft": 24,
              "paddingRight": 24,
              "paddingTop": 24,
              "width": "90%",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1A1A1A",
                "fontFamily": "SourceSansPro-Regular",
                "fontSize": 27,
                "marginBottom": 20,
              },
            ]
          }
        >
          Para saber mais sobre os projetos e instituiÃ§Ãµes que estamos ajudando, acesse o nosso site.
        </Text>
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "transparent",
              "borderColor": "#00843F",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 38,
              "justifyContent": "center",
              "marginBottom": 10,
              "opacity": 1,
              "width": 90,
            }
          }
          testID="button"
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#00843F",
                  "fontFamily": "SourceSansPro-SemiBold",
                  "fontSize": 27,
                },
              ]
            }
            textColor="primary"
          >
            Ir para o site
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
